### DISCUSSION TITLE
How to download a list of Participants team members

### DISCUSSION THREAD

**Discussion by emilymbender** [delph-in/docs#29] [Q&A]
Hi all,

I'm looking to download (as csv, etc) a single file with the list of people in the Participants team (as a step towards figuring out which DELPH-IN participants aren't there yet, so I can invite them). I have been unsuccessful in locating a way to do this within github. Found so far:

* The Teams API, which seems like way overkill. https://docs.github.com/en/rest/reference/teams
* These instructions, which suggest there should be a download CSV button which I can't find: https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/viewing-people-with-access-to-your-repository
* I can see the info here, but it's broken over multiple pages and also lists variable info for each person (though that might be al the info Github has): https://github.com/orgs/delph-in/teams/participants/members

For now, I'll do the clumsy cut & paste to get the info out of the third option, but surely there's a better way to do this?

Oh, and while I'm here griping about the GitHub UI, apparently I can select a team on this page, but then EXACTLY NO ACTIONS are available to me to do with the selected team: https://github.com/orgs/delph-in/teams

Thanks, I hate it here.
Emily

**Comment by goodmami** [comment]
Thanks for your patience as we try to find solutions for these kinds of actions. We weren't aware that it was an available or used feature of MoinMoin.

Unfortunately it looks like the "Export CSV" button is only available for enterprise users of GitHub and not the free tier. From further down on the [docs you linked](https://docs.github.com/en/enterprise-server@3.0/organizations/managing-access-to-your-organizations-repositories/viewing-people-with-access-to-your-repository#exporting-a-list-of-people-with-access-to-your-repository) (emphasis added):

> Owners of organizations *on GitHub Enterprise Cloud or GitHub Enterprise Server* can export a CSV list of people who have access to a repository.

Regarding the info at https://github.com/orgs/delph-in/teams, it only lists membership, but you can filter by role (e.g., maintainer, member). I agree it would be more useful if it showed the team role in a second column. You can also get more info about all delph-in organization members, such as whether 2FA is enabled, whether they have made their membership public, and their role in the organization, here: https://github.com/orgs/delph-in/people

When you select a team on the teams page, you should get some actions on a drop-down that appears next to the "select-all" check box, at the top of the list, but the only actions I see are to change the team's visibility or to delete the team.

If you need a CSV of participants and aren't an enterprise user, I think the best option is, unfortunately, through the API. GitHub has an [official command-line interface](https://cli.github.com/) through which you can send API requests. I'll see if I can get something working and report back.

**Reply by goodmami** [reply]
Ok, here's a solution, even if it's not very user-friendly. The REST API gives you a bunch of info you don't need and you need to execute multiple queries to assemble all the interesting info. Instead you can use the [GraphQL API](https://docs.github.com/en/graphql) to pull only the relevant info from multiple places. I recommend using the [gh](https://cli.github.com/) command line program for these because it will help you with authentication (which you need to get privileged info like membership rosters). This means you'll need to authorize the app for access to GitHub authenticated as you.

The query looks like this (adjust as necessary; this one is hard-coded for the "participants" team in the "delph-in" organization):

```console
$ gh api graphql -f query='
query {
  organization (login: "delph-in") {
    team (slug: "participants") {
      members {
        edges {
          node {
            login
            name
          }
          role
        }
      }
    }
  }
}' > participants.json
```

You may want to check that file to make sure it has the expected contents. If not, the next steps won't work.

To get a tab-separated file, you can use a Python script like this:

```python
# file: gh2tsv.py
import sys, json, csv
data = json.load(sys.stdin)
members = data['data']['organization']['team']['members']['edges']
with open('participants.tab', 'w', newline='') as csvfile:
    writer = csv.writer(csvfile, dialect='excel-tab')
    writer.writerow(['username', 'name', 'role'])
    for member_edge in members:
        writer.writerow([member_edge['node']['login'],
                         member_edge['node']['name'],
                         member_edge['role']])
```

This script only works for the query above because GraphQL responses change shape depending on what the query requests, but you can use it like this:

```console
$ cat participants.json | python3 gh2tsv.py
```

It will produce `participants.tab` when it is complete. You'll probably want to wrap the above in a single script so you don't need to copy-paste the commands each time.

**Reply by emilymbender** [reply]
Thanks, Michael and thanks for putting up with my grumpiness.

For posterity: MoinMoin didn't have this functionality per se. Rather, the relevant list of people was just stored as the wiki page ParticipantsGroup from which I could copy it.

**Reply by goodmami** [reply]
I see, and that one wiki means you don't have to click through multiple pages to copy/paste them all.

Speaking of which, I didn't include pagination limits in the query above, but there are [limits](https://docs.github.com/en/graphql/overview/resource-limitations). That is, I think the above wouldn't capture everyone if there were > 100 participants. If you use the solution above, I can modify it to check for pagination and retrieve successive pages.

**Reply by emilymbender** [reply]
Thanks, Michael. For this time, I think I'll probably just go ahead with the awkward copy paste --- unless there's more info in the json that you can export that way? It honestly sounds like way more work to get gh set up etc, even after your kind assistance with figuring out the query.

**Reply by goodmami** [reply]
What you can get out of `gh` depends on what permissions you grant it and how complex you make the query. For instance, I was unable to get user email addresses because I didn't grant it some permissions to do so. But, yes, in theory you could get a bit more info, at least what the users have made public.

To set up `gh` (presumably on a mac) it's just [brew install gh](https://github.com/cli/cli#installation) and [gh auth login](https://cli.github.com/manual/). IIRC GitHub will prompt you for what permissions you wish to grant the tool.

**Reply by goodmami** [reply]
@emilymbender good news: the push-button export of members is now available for regular (non-enterprise) GitHub organizations: https://github.blog/changelog/2021-10-06-export-organization-members/

I tested it out and the CSV option gives the following information:
1. Username
2. Public name
3. Whether 2FA is enabled
4. Whether the user's organization membership is public

(3) and (4) are only visible to owners and only owners can export the membership roster.

**Reply by emilymbender** [reply]
That is great news -- thanks!
